### PR TITLE
Custom Post Processes

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
Custom Post Processes

---
### Release Notes

To add the a new custom post process you need to use the context menu to create a new `Custom Post Process Volume C#`:

image

here is the C# code that will be generated (C#):
```csharp
using UnityEngine;
using UnityEngine.Rendering;
using UnityEngine.Rendering.HighDefinition;
using System;

[Serializable, VolumeComponentMenu("Post-processing/Custom/#NAME#")]
public sealed class #SCRIPTNAME# : CustomPostProcessVolumeComponent, IPostProcessComponent
{
    [Tooltip("Controls the intensity of the effect.")]
    public ClampedFloatParameter intensity = new ClampedFloatParameter(0f, 0f, 1f);

    Material m_Material;

    public bool IsActive() => m_Material != null && intensity.value > 0f;

    public override CustomPostProcessInjectionPoint injectionPoint => CustomPostProcessInjectionPoint.AfterPostProcess;

    public override void Setup()
    {
        if (Shader.Find("Hidden/Shader/#SCRIPTNAME#") != null)
            m_Material = new Material(Shader.Find("Hidden/Shader/#SCRIPTNAME#"));
    }

    public override void Render(CommandBuffer cmd, HDCamera camera, RTHandle source, RTHandle destination)
    {
        if (m_Material == null)
            return;

        m_Material.SetFloat("_Intensity", intensity.value);
        m_Material.SetTexture("_InputTexture", source);
        CoreUtils.DrawFullScreen(cmd, m_Material);
    }

    public override void Cleanup()
    {
        CoreUtils.Destroy(m_Material);
    }
}
```

Associated shader template:
```hlsl
Shader "Hidden/Shader/#SCRIPTNAME#"
{
    HLSLINCLUDE

    #pragma target 4.5
    #pragma only_renderers d3d11 ps4 xboxone vulkan metal switch

    #include "Packages/com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl"
    #include "Packages/com.unity.render-pipelines.core/ShaderLibrary/Color.hlsl"
    #include "Packages/com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl"
    #include "Packages/com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/FXAA.hlsl"
    #include "Packages/com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/RTUpscale.hlsl"

    struct Attributes
    {
        uint vertexID : SV_VertexID;
        UNITY_VERTEX_INPUT_INSTANCE_ID
    };

    struct Varyings
    {
        float4 positionCS : SV_POSITION;
        float2 texcoord   : TEXCOORD0;
        UNITY_VERTEX_OUTPUT_STEREO
    };

    Varyings Vert(Attributes input)
    {
        Varyings output;
        UNITY_SETUP_INSTANCE_ID(input);
        UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(output);
        output.positionCS = GetFullScreenTriangleVertexPosition(input.vertexID);
        output.texcoord = GetFullScreenTriangleTexCoord(input.vertexID);
        return output;
    }

    // List of properties to control your post process effect
    float _Intensity;
    TEXTURE2D_X(_InputTexture);

    float4 CustomPostProcess(Varyings input) : SV_Target
    {
        UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input);

        uint2 positionSS = input.texcoord * _ScreenSize.xy;
        float3 outColor = LOAD_TEXTURE2D_X(_InputTexture, positionSS).xyz;

        return float4(outColor, 1);
    }

    ENDHLSL

    SubShader
    {
        Pass
        {
            Name "#NAME#"

            ZWrite Off
            ZTest Always
            Blend SrcAlpha OneMinusSrcAlpha
            Cull Off

            HLSLPROGRAM
                #pragma fragment CustomPostProcess
                #pragma vertex Vert
            ENDHLSL
        }
    }
    Fallback Off
}
```

By default with the volume system you will be able to add you custom post process into a volume component: 
image

When you're done writing your custom post process you need to add it to the list of custom post processes in the project. This list allow you to organize the execution order of your custom post processes, as well as if they're executed or not:
image

Added graphic test: it tests the 3 injection points:
+ On before transparent, the GreenPP convert the luminance of the image to green
+ On Before post process the RedPP reads the red and green channels and writes to the green and red channels respectively 
+ On After post process the BluePP reads the red and green channels and output to the blue and red channels respectively

So the result is this blue image with a red transparent quad in the middle:
image

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/HDRP%252Ffeature%252Fcustom_post_process

---
### Overall Product Risks
**Technical Risk**: Medium

**Halo Effect**: Low

### CODE REVIEW

**Comment by sebastienlagarde** [comment]
Looks good to me. Waiting for Thomas reply.

**Comment by Chman** [comment]
Also please add graphic tests :)

File: com.unity.render-pipelines.high-definition/Editor/PostProcessing/Templates/CustomPostProcessingVolume.template [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    public override CustomPostProcessInjectionPoint injectionPoint => CustomPostProcessInjectionPoint.AfterPostProcess;
+
+    public override void Setup()
+    {
+        if (Shader.Find("Hidden/Shader/#SCRIPTNAME#") != null)
+            m_Material = new Material(Shader.Find("Hidden/Shader/#SCRIPTNAME#"));
+    }
+
+    public override void Render(CommandBuffer cmd, HDCamera camera, RTHandle source, RTHandle destination)
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/PostProcessing/Templates/CustomPostProcessingVolume.template]: Fixed using HDUtils.DrawFullScreen(cmd, m_Material, destination);

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Settings/DefaultSettingsPanel.cs [hunk 2]
```diff
@@ -161,6 +184,74 @@ void Draw_DefaultFrameSettings()
...
+
+                    reorderableList = new ReorderableList(customPostProcessTypes, typeof(string));
+                    reorderableList.drawHeaderCallback = (rect) =>
+                        EditorGUI.LabelField(rect, headerName, EditorStyles.boldLabel);
+                    reorderableList.drawElementCallback = (rect, index, isActive, isFocused) =>
+                    {
+                        var elemType = Type.GetType(customPostProcessTypes[index]);
+                        rect.height = EditorGUIUtility.singleLineHeight;
+
+                        EditorGUI.LabelField(rect, $"Order: {index}");
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Settings/DefaultSettingsPanel.cs]: I removed it

File: com.unity.render-pipelines.high-definition/Runtime/PostProcessing/PostProcessSystem.cs [hunk 3]
```diff
@@ -240,6 +241,16 @@ public void Cleanup()
             m_BokehIndirectCmd          = null;
             m_NearBokehTileList         = null;
             m_FarBokehTileList          = null;
+
+            // Cleanup Custom Post Process
+            if (HDRenderPipeline.defaultAsset != null)
+            {
+                foreach (var typeString in HDRenderPipeline.defaultAsset.beforePostProcessCustomPostProcesses)
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/PostProcessSystem.cs]: Fixed

**Pull request merged**